Simpy compatibility layer

maxfischer2781 opened the pull request:
Allow running SimPy simulations inside uSim. This pull request includes:

* the ``usim.py`` package, a drop-in replacement for ``simpy``,
* basic SimPy functionality of ``Environment`` and ``Event``s,
* the ability to run ``simpy`` standalone or embedded into ``usim``,
* the ability for uSim activities to ``await`` SimPy events,
* documentation including migration advice.

This pull request does *not* include:

* emulation of ``simpy.util`` and ``simpy.resources`` (to be added by another PR),
* the ability for SimPy processes to yield uSim notifications (to be added by another PR),
* emulation of ``simpy.rt`` real-time support.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mainekuehn/usim/25)
<!-- Reviewable:end -->

pep8speaks commented:
Hello @maxfischer2781! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`docs/conf.py`](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/docs/conf.py):

> [Line 21:1](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/docs/conf.py#L21): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file













* In the file [`usim_pytest/test_usimpy/test_events.py`](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py):

> [Line 216:13](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py#L216): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 217:13](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py#L217): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 218:13](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py#L218): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 219:13](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py#L219): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 229:13](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py#L229): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 230:13](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py#L230): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 231:13](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py#L231): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 232:13](https://github.com/MaineKuehn/usim/blob/3ef39a78b1d268518352c322514fa7a501334f49/usim_pytest/test_usimpy/test_events.py#L232): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+
+    def trigger(self, event: 'Event') -> 'Event':
+        """
+        Trigger this event to match the state of ``event``
```
What do you mean with *match the state of event*?

maxfischer2781 replied:
In SimPy, ``self`` copies the result and several flags from ``event``. In ``usim.py``, this is all folded into the result. How about this?

```suggestion
        Trigger this event and inherit the value or exception of ``event``
```

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+    unconditionally raises an error.
+    """
+    def __init__(self, *args, **kwargs):
+        raise NotEmulatedError(self.__doc__)
```
Should be adapted.

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+    process to run while others wait. Each :py:class:`~.Process` acts as an
+    event which is automatically triggered when the underlying generator completes.
+
+    .. code:: python3
```
I am afraid the examples don't speak for themselves. What do you want to show? E.g. what is the purpose of `clocks`? Is it only to show the difference to the `multi_clock` function below? Then I wouldn't get the point when reading until here.

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+        .. code:: python3
+
+            async def double_clock(sound):
+                "A clock of double duration"
```
This cannot be copied to try out the example. You could use a common comment here, for example.

maxfischer2781 replied:
```suggestion
                '''A clock of double duration'''
```

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+
+        .. code:: python3
+
+            async def clock(sound):
```
Is it on purpose to have two clock functions? The example below uses this clock function and not the other above. So by just copying the code users might run into problems.

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+        .. code:: python3
+
+            async def multi_clock():
+                "A clock of making multiple sounds"
```
```suggestion
                "A clock making multiple sounds"
```

Further you should again adapt the comment.

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+                break
+
+    @property
+    def is_alive(self):
```
This method might be better suitable in the `Event` itself.

maxfischer2781 replied:
Agreed, but the SimPy API is like this. I can implement it via the corresponding Event method, though.

maxfischer2781 replied:
Turns out ``Event`` doesn't have anything like this. The closes expression would be ``not self.triggered``, but that seems equally unhelpful. I recommend keeping it like this, any objections?

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+
+class ConditionValue:
+    """
+    ``dict``-like view to the events used in a condition and their value
```
Do they all share one value or do they have a value each?

```suggestion
    ``dict``-like view to the events used in a condition and their values
```

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+    .. note::
+
+        This type only captures the events, not their values.
+        If the value of an event changes, this class reflects the change.
```
I think the class itself cannot reflect the change but the instance of the class. How correct do you want to be?

maxfischer2781 replied:
How about this?

```suggestion
        If the value of a contained event changes, the value in the ConditionValue changes as well.
```

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+    .. code:: python3
+
+        def most(events, count):
+            "Test that at least 50% of events have triggered"
```
Again, the comment issue.

maxfischer2781 replied:
```suggestion
            '''Test that at least 50% of events have triggered'''
```

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+                else:
+                    event.defused = True
+                    self.fail(event.value)
+                    return
```
For this branch you could have a small function as you are using it in two places, on the top for observed and here.

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+                    continue
+                elif event.ok:
+                    unobserved.remove(event)
+                    observed.append(event)
```
From this point on you could think about only having a count for `observed` events as you are not accessing the events again. So you could introduce a local variable, initialise before while with `len(observed)` and then only do the increment.

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+        for event in events:
+            if isinstance(event, Condition):
+                result.extend(cls._flatten_values(event._events))
+            elif event.ok:
```
Why is here the distinction between `Condition` and `event.ok`, what about `event.fail`? (I suppose a failure is also a kind of value)

maxfischer2781 replied:
For a Condition, we *must* recurse to flatten it - this is because ``|`` and ``&`` produce a tree, not a chain. (See above comment why my attempts to fix this failed miserably.) If any event is ``event.fail``, then the entire Condition has failed already.

I guess an ``assert`` may be appropriate, both for documentation and robustness.

maxfischer2781 replied:
So I tried to come up with *some* paper assertion here but I did not find one. The publicAPI of ``Event`` is surprisingly bare. What could work is this:

```python3
assert event._value is None or event._value[1] is None or event.defused
```
Not sure if that is helpful at this point.

eileen-kuehn commented on usim/py/exceptions.py:
```diff
@@ -0,0 +1,40 @@
+This module defines the exceptions exclusive to the SimPy compatibility layer.
+These are used as shutdown signals between the environment and processes.
+Note that with the exception of :py:exc:`~usim.py.exceptions.Interrupt`,
+all signals should be considered internal and not meant to be handled manually.
```
```suggestion
all signals are considered internal and are not meant to be handled manually.
```

eileen-kuehn commented on usim/py/exceptions.py:
```diff
@@ -0,0 +1,40 @@
+"""
+
+
+class NotEmulatedError(NotImplementedError):
```
This Error we should consider for renaming. As mentioned in several other comments, the compatibility layer does not do a proper emulation but a simulation. However, talking about simulation error inside a simulation might be confusing for a user. So we should think about a proper name.

eileen-kuehn commented on usim/py/exceptions.py:
```diff
@@ -0,0 +1,40 @@
+
+
+class NotEmulatedError(NotImplementedError):
+    """An operation of the 'simpy' API is not emulated by the 'usim.py' API"""
```
```suggestion
    """An operation of the 'simpy' API is not mapped by the 'usim.py' API"""
```

eileen-kuehn commented on usim/py/exceptions.py:
```diff
@@ -0,0 +1,40 @@
+
+
+class SimPyException(Exception):
+    """Base case for all exceptions of ``usim.py`` that can safely be handled"""
```
What happens to exceptions from SimPy that cannot be handled safely?

maxfischer2781 replied:
The ``StopSimulation`` (and ``NotEmulatedError ``) is not supposed to be handled manually. How bout this?
```suggestion
    """Base case for all exceptions of ``usim.py`` that can safely be handled in a simulation"""
```

eileen-kuehn replied:
You could also say *during a simulation*. Change as you prefer :)

eileen-kuehn commented on usim/py/exceptions.py:
```diff
@@ -0,0 +1,40 @@
+
+    .. warning::
+
+        This exceptions exists for historical compatibility only.
```
```suggestion
        This exception exists for historical compatibility only.
```

eileen-kuehn commented on usim_pytest/test_usimpy/test_core.py:
```diff
@@ -0,0 +1,79 @@
+    def test_until_past(self, env):
+        """env.run(until) argument must not be in the past"""
+        with pytest.raises(ValueError):
+            env.run(-5)
```
Consider also testing edge cases, such as `0`. I think I put a comment somewhere, that `0` should not raise a `ValueError`.

maxfischer2781 commented on docs/source/topics/simpy.rst:
```diff
@@ -0,0 +1,103 @@
+
+To allow running, integrating and migrating existing SimPy simulations,
+μSim has an inbuilt compatibility layer: the :py:mod:`usim.py` module
+emulates most of the SimPy API. The only truly unsupported feature are scheduling
```
We could go with one of "imitate" or "replicate". What do you think?

eileen-kuehn replied:
In this case we might even go with *map* as I proposed in one of the other docstrings.

maxfischer2781 commented on docs/source/topics/simpy.rst:
```diff
@@ -0,0 +1,103 @@
+    ...         trip_duration = 2
+    ...         yield env.timeout(trip_duration)
+    ...
+    >>> import usim.py as simpy  # import usim.py instead of simpy
```
The SimPy example consists of two separate code snippets - the second one starts at the import. So it replicates the example, but is not exakt either way.

We could split this back up into two parts, showing that the first doesn't need any modification and the second one just the changed import. Sounds good?

maxfischer2781 replied:
I've split this example now. Makes the change much clearer.

maxfischer2781 commented on usim/py/core.py:
```diff
@@ -0,0 +1,319 @@
+from .. import Scope
+
+from .events import Event
+from .exceptions import NotEmulatedError, StopSimulation, StopProcess
```
This is related to not saying that we emulate simpy?

maxfischer2781 commented on usim/py/core.py:
```diff
@@ -0,0 +1,319 @@
+            raise NotEmulatedError(
+                "'env.run' is not supported inside a 'usim' simulation\n"
+                "\n"
+                "Synchronous running blocks the event loop. Use instead\n"
```
```suggestion
                "Synchronous 'run' block the event loop. Use instead:\n"
```
Matching wording of similar error messages.

maxfischer2781 commented on usim/py/core.py:
```diff
@@ -0,0 +1,319 @@
+                "* 'await env.until()' to block only the current activity\n"
+                "* 'async with env:' to concurrently run the environment\n"
+                "\n"
+                "You can 'env.run' outside of a 'usim' simulation"
```
How about "may"?

maxfischer2781 commented on usim/py/core.py:
```diff
@@ -0,0 +1,319 @@
+
+            This method exists for historical compatibility only.
+            Use ``return value`` instead.
+        """
```
Older Python versions did not allow ``return`` in a generator. So SimPy processes could not ``return`` a value either. This really only exists for completeness. Should this be made clearer? I guess it *is* confusing without in-depth knowledge.

maxfischer2781 commented on usim/py/core.py:
```diff
@@ -0,0 +1,319 @@
+        The μSim compatibility layer uses the regular μSim event loop.
+        There is no public alternative to 'Environment.step'.
+        """
+        raise NotEmulatedError(self.step.__doc__)
```
Any suggestion how? ;)

maxfischer2781 commented on usim/py/events.py:
```diff
@@ -0,0 +1,642 @@
+    # Implementation Note
+    # ``usim`` would flatten Conditions of the same type here. With the
+    # ``simpy`` model, this does not work well. The problem is that events
+    # are active components that may fail the Environment.
```
I am willing to be less-than exact in these comments - they are hopefully only relevant for people that murk around in the internals anyways. In this case, it should make sense at least in hindsight.

maxfischer2781 commented on usim/py/core.py:
```diff
@@ -0,0 +1,319 @@
+                if isinstance(until, Event):
+                    await until.__usimpy_flag__
+                else:
+                    if until <= time.now:
```
Allowed ``until == time.now`` and added unit test.

maxfischer2781 commented:
I've either directly committed or implemented most things addressed. We should decide on a better word for "emulate" and try and come up with an example for ``Process``.

maxfischer2781 commented:
Renamed ``NotEmulatedError`` to a more specific ``NotCompatibleError`` for things we *cannot* (reasonably) provide -- things we just did not do are a regular ``NotImplementedError``. Where appropriate, replaced "emulate" with alternatives such as "provide" or "recreate".

maxfischer2781 commented:
I've swapped the Process example from clocks to travelling.

eileen-kuehn commented on usim/py/events.py:
```diff
@@ -0,0 +1,658 @@
+            async def visit_many(*destinations: str):
+                '''Visit all destinations one after the other'''
+                for destination in destinations:
+                    # awat a bare activity to wait for its completion
```
```suggestion
                    # await a bare activity to wait for its completion
```

maxfischer2781 commented on usim/py/__init__.py:
```diff
@@ -0,0 +1,126 @@
+Resources
+---------
+
+``usim.py`` implements no resources yet.
```
Not the ``simpy`` resources. Since they are not needed to get the compatibility layer running, I would prefer adding them in a later commit.

maxfischer2781 commented on usim/py/core.py:
```diff
@@ -0,0 +1,321 @@
+
+
+class Environment:
+    """
```
Needed to allow ``\ `` escape.

```suggestion
    r"""
```
